juju/juju
# Add support for k8s operators behind a "k8s-operators" feature flag.

## Checklist

 - [x ] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?

Controller only facade changes.

## Description of change

An operator charm is one with deployment mode = "operator".
It does not call k8s-spec-set - instead the charm pod is the workload.

The main changes are to the caas unit provisioner worker and k8s broker.
The worker does not start a deployment worker anymore as that deployment worker is only used to scale and handle k8s spec changes - both of these are disabled for operators. It also tweaks how it watches the cluster to report juju application and unit info - it watches the charm operator. The worker uses a new DeployentMode api call to see what type of charm it is.

The relevant k8s broker methods take a mode argument (operator or workload) and tweak the name of the resources they look for and/or the label selector used.

## QA steps

bootstrap a k8s controller, enable the feature flag
`$ juju controller-config features=[k8s-operators]`

deploy a charm with metadata:

```yaml
deployment:
  mode: operator
```
and which does not call k8s-spec-set.

juju status will show the app and a single unit as active
juju show-status-log will show all the start up hooks run
juju scale-application will result in an error
kubectl get all will show only the operator pod etc in the cluster

Check that mariadb-k8s still deploys and scales as before.
